### PR TITLE
Version 92

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

**Comment by sbc100** [comment]
What should go in the change log for this revision range?

**Comment by sbc100** [comment]
I'm mostly want to create this release as the last one before I switch to github actions, so we can compare the resulting release with the one github actions produces apples-to-apples.

**Comment by sbc100** [comment]
Lets fill it in now.  Here is some inspiration:

```
a308fc323 Fix issues with Types and Features (#2773) 25 hours ago [Thomas Lively]
e30f52e59 Report errors in scripts/clang-tidy-diff.sh if clang-tidy is not found (#2775) 26 hours ago [Sam Clegg]
727c91af3 Dummy interpreter support for EH (#2774) 26 hours ago [Heejin Ahn]
c9162258a Disable multivalue in fuzzer in a clearer way (#2771) 32 hours ago [Alon Zakai]
0d5c9dc7b Validate that tuples have multiple operands (#2768) 33 hours ago [Thomas Lively]
5177b9180 Fix OOB fuzzing (#2769) 34 hours ago [Alon Zakai]
2302955b1 Emit tuples in the fuzzer (#2695) 2 days ago [Thomas Lively]
598b3dffa Enable cross-VM fuzzing + related improvements to fuzz_opt.py (#2762) 2 days ago [Alon Zakai]
e04d25e2e Fix reuse of constant nodes in Precompute (#2764) 3 days ago [Heejin Ahn]
359525bc5 Note removal of catch body in Vacuum (#2765) 3 days ago [Heejin Ahn]
ec76e9e89 Fix build with a sprinkling of braces (#2766) 3 days ago [Thomas Lively]
9ee270633 Fix auto updater on spec tests after #2755 (#2763) 4 days ago [Alon Zakai]
1bddd6be5 Fix Atomics fuzz bugs in interpreter (#2760) 4 days ago [Alon Zakai]
c16bfeebb Use direct pointers as Type IDs (#2745) 4 days ago [Thomas Lively]
c45ae1649 Update v8 flags (#2754) 4 days ago [Alon Zakai]
e0716f533 Add --deterministic flag to wasm2js, for fuzzing (#2757) 4 days ago [Alon Zakai]
585f8174c Fix invalid index checking in readImports (#2758) 4 days ago [Heejin Ahn]
13efc1eca Speed up wasm-reduce test (#2756) 4 days ago [Alon Zakai]
d7d251b75 Use std::cout for interpreter trap logging (#2755) 5 days ago [Alon Zakai]
53c70d9cc Fuzz fix for SimplifyGlobals nopping (#2750) 5 days ago [Alon Zakai]
69aa3893b Remove duplicate Type:: prefixes (NFC) (#2753) 5 days ago [Heejin Ahn]
3f3e17788 Legalize return_calls properly (#2752) 5 days ago [Alon Zakai]
6d4974135 Fix multivalue event fuzzing (#2748) 7 days ago [Thomas Lively]
7126b2c70 Remove redundant vacume pass.  Followup on #2741 (#2747) 7 days ago [Sam Clegg]
6a51af5d8 Remove writes to globals that are never written to (#2741) 8 days ago [Sam Clegg]
1b56c2b34 Optimize uint32_t Type constructor (#2744) 8 days ago [Thomas Lively]
c27600385 Handle tuples in Asyncify call support (#2743) 8 days ago [Thomas Lively]
1cfaf2e72 [NFC] Update readme to discuss tuples rather than push/pop (#2722) 8 days ago [Thomas Lively]
e4ae33c93 Remove old setup.cfg (#2736) 8 days ago [Sam Clegg]
9c75b8f8e Remove function index printing (#2742) 8 days ago [Thomas Lively]
b3d79f406 Performance optimizations for Type (#2733) 8 days ago [Thomas Lively]
0dcb68967 Move scripts in the `scripts` directory (#2735) 8 days ago [Sam Clegg]
9d572bd24 Remove executable bit from test scripts (#2734) 8 days ago [Sam Clegg]
88b2b6411 Enable color diagnostics with ninja+gcc (#2739) 8 days ago [Sam Clegg]
f5df71a54 Factor out wasm_opt testing. NFC. (#2737) 8 days ago [Sam Clegg]
9ee5f403d Fix an old misleading comment (#2738) [ci skip] 8 days ago [Alon Zakai]
d81236ad8 Remove references to vanilla emcc (#2731) 9 days ago [Sam Clegg]
dc5a503c4 Remove .gitmodules file (#2732) 9 days ago [Sam Clegg]
2f9d01926 Avoid g$ in main modules where possible (#2721) 9 days ago [Alon Zakai]
95a66b5b3 Run reorder-locals more in wasm2js (#2729) 9 days ago [Alon Zakai]
0a44e995d Fix ReorderLocals handling of local names (#2728) 9 days ago [Alon Zakai]
7531e7b46 Work around a compiler issue on MacOS (#2730) 10 days ago [Alon Zakai]
c8a1489f2 Optimize ReorderLocals maps to vectors (#2727) 10 days ago [Alon Zakai]
e77d85fec Only do fp$ optimization in the main module (#2720) 10 days ago [Alon Zakai]
eaa724010 JS/Wasm BigInt support for wasm-emscripten-finalize (#2726) 10 days ago [Alon Zakai]
ac8b0b712 Do not emit multivalue events in fuzzer (#2723) 2 weeks ago [Thomas Lively]
80c5164e1 Tuple globals (#2718) 2 weeks ago [Thomas Lively]
702b99508 Revert to using globals in Asyncify call support (#2719) 2 weeks ago [Thomas Lively]
a32102c7d Avoid unnecessary fp$ in side modules (#2717) 2 weeks ago [Alon Zakai]
d8179402b Represent dylink section in IR, so we can update it. (#2715) 3 weeks ago [Alon Zakai]
2b758fbdc Error out when EH is used in unsupported code (#2713) 3 weeks ago [Heejin Ahn]
d595a45b6 Avoid fp$ access in MAIN_MODULES (#2704) 3 weeks ago [Alon Zakai]
9101f65c7 Tuple operations in C and JS APIs (#2711) 3 weeks ago [Thomas Lively]
ce6ae4986 Use Emscripten upstream when compiling binaryen.js (#2488) 3 weeks ago [Daniel Wirtz]
91061dace Add TAGS to .gitignore (#2710) 3 weeks ago [Thomas Lively]
5ddda8d2e Convert Asyncify fake globals to fake calls. NFC (#2706) 3 weeks ago [Alon Zakai]
e01e7e9db Disallow tuples in MVP (#2707) 3 weeks ago [Thomas Lively]
e7315448c Fix Event section ordering (#2708) 3 weeks ago [Thomas Lively]
18fcad6df Allow subtypes in tuple operations (#2700) 4 weeks ago [Thomas Lively]
03ae7fcbf SIMD integer abs and bitmask instructions (#2703) 4 weeks ago [Thomas Lively]
39fda77da Emit unreachable tuple.make properly (#2701) 4 weeks ago [Thomas Lively]
5946b9aa7 Support tuple locals in Asyncify (#2696) 4 weeks ago [Thomas Lively]
b4a8ace5c Dump initial wasts in fuzzer (#2697) 4 weeks ago [Thomas Lively]
595d8a3f3 makeConstExpression => makeConstantExpression (#2698) 4 weeks ago [Alon Zakai]
38a795c48 Fix binary emitting of signature indices (#2694) 5 weeks ago [Thomas Lively]
83d69b30b Handle tuples in RemoveUnusedBrs (#2693) 5 weeks ago [Thomas Lively]
aadec4526 Collect signatures from all block kinds (#2691) 5 weeks ago [Thomas Lively]
2b8e92486 Interpret tuple locals and tail-calls correctly (#2690) 5 weeks ago [Thomas Lively]
69d34c6bd Handle tuples in wasm-reduce (#2689) 5 weeks ago [Thomas Lively]
a3bbe87a9 Handle unreachable in TupleExtract::finalize (#2692) 5 weeks ago [Thomas Lively]
5c0bf993b Update RedundantSetElimination to work with tuples (#2688) 5 weeks ago [Thomas Lively]
0abc9ce8e Update Precompute to handle tuples (#2687) 5 weeks ago [Thomas Lively]
8f16059d3 Handle multivalue returns in the interpreter (#2684) 5 weeks ago [Thomas Lively]
49e31f203 Use emcmake on CI (#2686) 5 weeks ago [Alon Zakai]
5d7248b16 Add a non-const iterator to SmallVector (#2685) 5 weeks ago [Thomas Lively]
9bcc431a4 Add 'warning:' to names section warning. Helps #2680 (#2683) 6 weeks ago [Alon Zakai]
dc264342e Asyncify: Fix wasm-only instrumentation of unnamed imports (#2682) 6 weeks ago [Alon Zakai]
9e3dbb1f6 Initial multivalue support (#2675) 6 weeks ago [Thomas Lively]
3a275d062 Update build instructions in README (#2667) 6 weeks ago [Thomas Lively]
5034bf1dd DWARF: Ignore a compile unit with no abbreviations (#2678) 6 weeks ago [Alon Zakai]
57a81a04f Expose asyncify state via a getter (#2679) 6 weeks ago [Alon Zakai]
2a0f6a513 Improve a comment on unreachability [ci skip] (#2672) 6 weeks ago [Alon Zakai]
6371cf636 Simplify binary block parsing (#2674) 7 weeks ago [Thomas Lively]
1f682f432 Generalize binary writing for tuples (#2670) 7 weeks ago [Thomas Lively]
f0834cc1a On OpenBSD (6.6) libc++ fileno is defined as a MACRO which doesn't work with :: (#2669) 7 weeks ago [osen]
210521497 Add multivalue feature (#2668) 7 weeks ago [Thomas Lively]
f79faeb14 Add EH support for CodeFolding (#2665) 7 weeks ago [Heejin Ahn]
b17e84b49 Fix to https://github.com/WebAssembly/binaryen/issues/2170 (#2654) 8 weeks ago [juj]
```

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,10 @@ full changeset diff at the end of each section.
 Current Trunk
 -------------
 
+v92
+---
+
+
```
Comment (tlively) [inline CHANGELOG.md:21]:
```suggestion
 - The `multivalue` feature has been added. It allows functions and control flow structures to return tuples and for locals and globals to have tuple types. Tuples are created with the new `tuple.make` pseudoinstruction and their elements are retrieved with the new `tuple.extract` pseudoinstruction.
 
```

**Comment by sbc100** [comment]
Anything from you alon?

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,16 @@ full changeset diff at the end of each section.
...
+v92
+---
+ - The `multivalue` feature has been added. It allows functions and control flow
+   structures to return tuples and for locals and globals to have tuple types.
+   Tuples are created with the new `tuple.make` pseudoinstruction and their
+   elements are retrieved with the new `tuple.extract` pseudoinstruction.
+ - The internal type interner has been rewritten to avoid taking locks in far
+   more situations. Depending on the workload, this may result in large speedups
+   and increased parallelism.
+
```
Comment (kripken) [inline CHANGELOG.md:27]:
```suggestion
- Represent the `dylink` section in Binaryen IR, so we can read, write,
  and update it.
```

**Comment by kripken** [comment]
Hmm, most of my stuff here is probably best mentioned (and is) in the emscripten changelog since it's very interconnected. But I added one suggestion now.

**Pull request merged**